### PR TITLE
allow individual plugins to load datapacks [wip]

### PR DESCRIPTION
**Pull request by ExcessiveAmountsOfZombies**
Tried making it so each plugin could also be its own datapack last night. Just create a folder named datapack in the plugin data folder, put some data in it and then it'll load onto the server [plugins](https://cdn.discordapp.com/attachments/308735693587349505/788595783531364362/unknown.png) seemed to continue working on /minecraft:reload.

it's draft because I didn't finish working on it, just made it functional and so the question I have is this something that makes sense to have? more of the game is being run through these datapacks and they can do quite a bit so I feel it makes sense to allow plugin to also be a datapack if it wants to be. This is probably something that would work well with https://github.com/PaperMC/Paper/pull/3625 assuming that machine went with hooking into the serialization system. 

If this is wanted what I would want to add is:
- ability to get the datapack folder,
- ability to change the datapack folder name so it's not just "datapack,"

The only current issue I noticed was that when you restart the server the plugin datapacks give a warning in the console because they are not loaded for the first pass (in the Main class when the game first populates the "vanilla/builtin" datapack)

### CODE REVIEW

File: Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+Subject: [PATCH] Allow individual plugins to load datapacks
+
+
+diff --git a/src/main/java/com/destroystokyo/paper/datapack/PluginDataPack.java b/src/main/java/com/destroystokyo/paper/datapack/PluginDataPack.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..ec8d5c32ea14bc4941a318d9b8177627950ebfa3
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/datapack/PluginDataPack.java
+@@ -0,0 +1,59 @@
++package com.destroystokyo.paper.datapack;
```
Comment (BillyGalbreath) [inline Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch]: Paper is moving to the io.papermc package

File: Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+ 
+diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
+index ae48df4b9b78bc3a687166fdaa7b8a60bd13ec35..a484f8e24772225812d26f411f8febfcd10f41fe 100644
+--- a/src/main/java/net/minecraft/server/Main.java
++++ b/src/main/java/net/minecraft/server/Main.java
+@@ -119,7 +119,6 @@ public class Main {
+             if (flag) {
+                 Main.LOGGER.warn("Safe mode active, only vanilla datapack will be loaded");
+             }
+-
```
Comment (BillyGalbreath) [inline Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch]: Unnecessary diff

**Review by BillyGalbreath — CHANGES_REQUESTED** [review]
Super hyped for this <3 \o/

**Comment by electronicboy** [comment]
Billy approval, *tosses in the trashbin*

`The only current issue I noticed was that when you restart the server the plugin datapacks give a warning in the console because they are not loaded for the first pass`
The issue for stuff like this is that you've gotta be able to either serialize them to the disk to be loaded (which causes its own set of woes), or, delay that phase until plugins have had a time to do that stuff, but, then you're also delaying work on a platform which slim chance may cause issues elsewhere which would need to be catered for, e.g. bukkit already has support for loading this stuff in the API last I recall

**Comment by ExcessiveAmountsOfZombies** [comment]
> The issue for stuff like this is that you've gotta be able to either serialize them to the disk to be loaded (which causes its own set of >woes), or, delay that phase until plugins have had a time to do that stuff, but, then you're also delaying work on a platform which > slim chance may cause issues elsewhere which would need to be catered for

It's really only an issue in that it's a warning in the console. You don't need your plugin datapack loaded in the Main class in NMS and it could be put off into the second pass in the DedicatedServer class. If that can be solved I don't think there would be any problems with per plugin datapacks.

I know the list of datapacks is stored in the world somewhere, probably level.dat so some ideas I had thought of:
Not saving the plugin datapacks names to the world
adding a check to make sure plugin datapacks don't get loaded until after plugins have been called onLoad

>  bukkit already has support for loading this stuff in the API last I recall

loading datapacks? I haven't seen anything but admittedly I still find things in the Bukkit API occasionally I didn't know existed.

File: Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch (hunk 1 above)
Comment (ExcessiveAmountsOfZombies) [reply Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch]: Thanks!

**Comment by theminecoder** [comment]
Regarding this, I think plugin based datapacks would be a little bit more intuitive and sharable if they could be loaded from the plugins jar itself via way of included resources.

With the current scenario plugins would need to essentially export the datapack from the jar into the datapack folder in the `onLoad` method which some plugins will probably do for customisation reasons but for private plugins and things that don't need to be customised this serves no purpose but to add extra code to do the exporting.

How you would handle priority for this I'm not too sure (plugin.yml might be an option maybe) but for adoptability I think having the option to have jar based loading is very important.

**Comment by ExcessiveAmountsOfZombies** [comment]
> Regarding this, I think plugin based datapacks would be a little bit more intuitive and sharable if they could be loaded from the plugins jar itself via way of included resources.

Ok, I just changed it to do this instead of having a datapack folder, I actually like this method of doing it more and agree that it makes more sense to do it this way.

This also lowers the priority of plugin based datapacks because the user doesn't really have an option to modify them now so world based datapacks now have a higher priority than plugin ones. I removed the world datapack loading in the first pass in Main.java because of a future assumption that plugins will be able to add more custom things that would require a plugin to be loaded first.

still has a few issues that keep this from being perfect though

1)
>[16:46:44 WARN]: Missing data pack entry/ThonkCrafting
[16:46:44 WARN]: Missing data pack file/bukkit
[16:46:44 WARN]: Missing data pack file/strider-s-redux-1-1-0.zip

warning on first datapack pass. doesn't affect anything but ideally it wouldn't cause a warning.

2) I looked at this a bit and it doesn't look like plugins actually disable if they fail the onLoad stage so the datapack for it will still load even if the plugin fails.

>How you would handle priority for this I'm not too sure (plugin.yml might be an option maybe) but for adoptability I think having the option to have jar based loading is very important.

I don't think priority between plugins matters that much, I don't think it's very likely that a user would install a plugin that would conflict with another one.

**Comment by Machine-Maker** [comment]
So an issue that was brought up is, do the datapacks need to be loaded *after* plugins? If all you are doing is reading the jar for a datapack, you can do that whenever, separate from the whole loading of plugins. @ExcessiveAmountsOfZombies mentioned you might want to do that after loading plugins, because the datapacks might reference some custom feature or recipe, or something that the plugin itself loads. I don't know 100% about how the whole loading of datapacks work, but it *might* be the case that it doesn't matter that the datapack references something that the server won't understand until later in starup, and if that is true, then you can just load the datapacks from the jars right where it takes place now.

**Comment by Machine-Maker** [comment]
From just a simple test just now, there aren't any issues with referencing a recipe in an advancement reward that the server doesn't know about until a plugin loads, so I think it might be fine to keep plugin loading where it is, and just load the datapack from the plugin where datapacks are normally loaded.

**Comment by ExcessiveAmountsOfZombies** [comment]
Ok, i've moved up the plugin datapacks into the Main.java to remove the error that occurs when it loads. I just look for the jars in the plugin folder and then load them appropriately. 

I still believe plugin loading will need to happen sooner than it currently does in the future, but since modifying/adding with NMS isn't supported(? it's not on spigot so I assume the same here) and there are currently no APIs to add custom serializers that datapacks can reference I guess it's not that big of a deal.

Example: https://gist.github.com/ExcessiveAmountsOfZombies/99a23547a893335cfac89b1d1fbca4b1

**Comment by Machine-Maker** [comment]
I still think it'd be better if the datapack was contained in a folder, instead of just having the `data` and `pack.mcmeta` folder in the root of the jar. From an organizational point of view, it just makes more sense. Otherwise you have a seemingly random data directory in there. Also, if mojang adds more root directories (other than `data` and `assets` itll be best to have all those contained in a folder)

EDIT: Also, I think the name of that folder should be configurable in `plugin.yml` defaulting to something like "datapack".

**Comment by Machine-Maker** [comment]
Also, having some sort of load order would be a good thing. When you enable datapack via the `/datapack` command, you can chose to enable it first, last, after a pack, or before a pack. Maybe something as simple as a integer in plugin.yml that determines the order of loading.

File: Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
++        this.pluginFolder = pluginFolder;
++    }
++
++    @Override
++    public void a(Consumer<ResourcePackLoader> consumer, ResourcePackLoader.a creator) {
++        List<PluginDataPack> packs = new ArrayList<>();
++
++        if (pluginFolder.exists()) {
++            for (File file : pluginFolder.listFiles((dir, name) -> name.endsWith(".jar"))) {
++                PluginDataPack dataPack = new PluginDataPack(FilenameUtils.removeExtension(file.getName()), file, true);
```
Comment (Machine-Maker) [inline Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch:71]: I think the name of the pack should really reflect the plugin name in `plugin.yml` rather than just the name of the jar.

File: Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
++                if (!dataPack.a(EnumResourcePackType.SERVER_DATA).isEmpty()) {
++                    packs.add(dataPack);
++                }
++
++            }
++        }
++
++        for (PluginDataPack pack : packs) {
++            ResourcePackLoader resourcePack = ResourcePackLoader.a("entry/" + pack.getName(), false,
```
Comment (Machine-Maker) [inline Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch:80]: I'd call it `plugin/<name>` rather than `entry/<name>`.

Also, OBFHELPER

File: Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
++                () -> pack, creator, ResourcePackLoader.Position.BOTTOM, PackSource.d);
```
Comment (Machine-Maker) [inline Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch:81]: OBFHELPER

File: Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
+index 172fc9ef9c0d3444eb99f750a17d42f130d94f73..0d94a049c1e2311ea40fbc7e13277abbc3b0e1df 100644
+--- a/src/main/java/net/minecraft/server/MinecraftServer.java
++++ b/src/main/java/net/minecraft/server/MinecraftServer.java
+@@ -132,7 +132,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
+     private long nextTick;
+     private long W; final long getTickOversleepMaxTime() { return this.W; } // Paper - OBFHELPER
+     private boolean X; final boolean hasExecutedTask() { return this.X; } // Paper - OBFHELPER
+-    private final ResourcePackRepository resourcePackRepository;
++    protected ResourcePackRepository resourcePackRepository; // paper - widen access
```
Comment (Machine-Maker) [inline Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch:120]: Is this change needed since you moved stuff into Main?

File: Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
++    public PluginSource(File pluginFolder) {
++        this.pluginFolder = pluginFolder;
++    }
++
++    @Override
++    public void a(Consumer<ResourcePackLoader> consumer, ResourcePackLoader.a creator) {
++        List<PluginDataPack> packs = new ArrayList<>();
++
++        if (pluginFolder.exists()) {
++            for (File file : pluginFolder.listFiles((dir, name) -> name.endsWith(".jar"))) {
```
Comment (Machine-Maker) [inline Spigot-Server-Patches/0616-Allow-individual-plugins-to-load-datapacks.patch:70]: Should every reload look for new files with the `.jar` name? I don't think new plugins' datapacks should be reloaded with /minecraft:reload if the server doesn't know about those plugins already. Maybe move this to constructor and store a list of PluginDataPacks.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
Do any of the API changes need to stay? I don't think you are using Plugin#getFile since you moved datapack loading before plugin loading.

File: Spigot-API-Patches/0242-extend-ResourcePackFile-to-use-jar-based-datapacks-a.patch [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+         return org.slf4j.LoggerFactory.getLogger(getLogger().getName());
+     }
++
++    /**
++     * Returns the file where the jar is located.
++     *
++     * @return The jar.
++     */
++    @NotNull
++    public File getFile();
```
Comment (kashike) [inline Spigot-API-Patches/0242-extend-ResourcePackFile-to-use-jar-based-datapacks-a.patch:24]: Two things:
- return `Path` instead
- call this `getJarPath`

**Review by kashike — CHANGES_REQUESTED** [review]

**Comment by kittech0** [comment]
year later and nothing...

**Comment by Owen1212055** [comment]
Closing PR, as it seems that this has gone inactive.

If you're still interested in continuing this PR, feel free to leave a comment and we can reopen it for you. 😄
